### PR TITLE
pawn logic and test 

### PR DESCRIPTION
**Pull request by leamerluck**
UPDATE 8/19/15: Ok, thanks to Stephen I included more tests that caught some bugs in my code. I do believe that with his help we have covered the possibilities for the pawn. Also, I did a rebase so all those commits are corrected and it looks a little cleaner :smile: 

UPDATE: This is ready for merging I believe. Sorry about all commits.. I forgot to change the location of  `obstacle?` It was passing all tests and then I pulled master down and forgot to run rake again. All tests are passing. Take a gander at the code and shoot me any questions. 

NOT ready for merging. I just wanted to throw this up here for everyone to take a gander. Still working on getting `is_obstructed?` integrated with the Pawn logic.

### CODE REVIEW

**Comment by hugolmendez** [comment]
:+1:

**Comment by hugolmendez** [comment]
:+1:

File: test/models/pawn_test.rb [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	test "check all methods together within valid_move?" do
+		# Test first move up 2 but x changes by 1
+		expected = false 
+		actual = @white_pawn.valid_move?(4,2)
+
+		assert_equal expected, actual
+
+		# Test 1st move 1 forward but met by obstacle
+		@black_pawn.update(:y_position => 4)
+		actual = @white_pawn.valid_move?(4,1)
```
Comment (CodeFlowed) [inline test/models/pawn_test.rb]: @leamerluck I haven't checked all the other tests yet; but on line 138, it seems like the arguments for valid_move are invalid (a pawn with location—1, 2—can't move to the right). Did you mean `valid_move?(1,4)`? 
And if that's the case, you're actually moving the white pawn by 2 squares, not 1. So you might need another test for when a pawn moves 1 square but is blocked.

File: app/models/pawn.rb [hunk 2]
```diff
@@ -1,3 +1,44 @@
...
+		[move_number_allowed?(new_y), no_vertical_obstacle?(new_y), not_a_backwards_move?(new_y), diagonal_move_allowed?(new_x, new_y)].all?
+	end
+
+	def move_number_allowed?(new_y)
+		move_number(new_y) == 1 || (move_number(new_y) == 2 && first_move?(new_y))
+	end
+
+	def move_number(new_y)
+		(new_y - y_position).abs
+	end
```
Comment (CodeFlowed) [inline app/models/pawn.rb]: This def could probably be replaced with a variable like `move_number = (new_y - y_position).abs` making the code a little faster and DRY-er.

File: app/models/pawn.rb [hunk 3]
```diff
@@ -1,3 +1,44 @@
...
+	  (color == 1 && y_position == 2) || (color == 0 && y_position == 7) 
+  end
+
+	def no_vertical_obstacle?(new_y)
+		move_number(new_y) == 2 ? !is_obstructed?(x_position, new_y) && no_vertical_adjacent_obstacle?(x_position, new_y) : (no_vertical_adjacent_obstacle?(x_position, new_y))
+	end
+
+	def no_vertical_adjacent_obstacle?(x_position, new_y)
+		!obstacle?(x_position, new_y)
+	end
```
Comment (CodeFlowed) [inline app/models/pawn.rb]: This def `no_vertical_adjacent_obstacle?` can probably be replaced with a variable like `no_vertical_adjacent_obstacle? = !obstacle?(x_position, new_y)`, too.

File: app/models/pawn.rb [hunk 4]
```diff
@@ -1,3 +1,44 @@
 class Pawn < Piece
 
+	def valid_move?(new_x, new_y)
+		[move_number_allowed?(new_y), no_vertical_obstacle?(new_y), not_a_backwards_move?(new_y), diagonal_move_allowed?(new_x, new_y)].all?
```
Comment (CodeFlowed) [inline app/models/pawn.rb]: Splitting line 4 into multiple lines like:
```
  def valid_move?(new_x, new_y)
    [move_number_allowed?(new_y),
     no_vertical_obstacle?(new_y),
     not_a_backwards_move?(new_y),
     diagonal_move_allowed?(new_x, new_y)].all?
  end
```
will make it a lot easier to read on Github or small screens (80 characters is the preferred line length according to the [Ruby style guide](https://github.com/styleguide/ruby)).

File: app/models/pawn.rb [hunk 5]
```diff
@@ -1,3 +1,44 @@
...
+	def move_number(new_y)
+		(new_y - y_position).abs
+	end
+
+	def first_move?(new_y)
+	  (color == 1 && y_position == 2) || (color == 0 && y_position == 7) 
+  end
+
+	def no_vertical_obstacle?(new_y)
+		move_number(new_y) == 2 ? !is_obstructed?(x_position, new_y) && no_vertical_adjacent_obstacle?(x_position, new_y) : (no_vertical_adjacent_obstacle?(x_position, new_y))
```
Comment (CodeFlowed) [inline app/models/pawn.rb]: Splitting lines 20 and 33 into multiple lines by making them normal `if` statements can make it more readable and again easier to view on smaller screens. Sorry about my nitpicking, Chris. 😅

**Comment by CodeFlowed** [comment]
@leamerluck The revamped code here is a huge improvement! Clever use of the `all?` method in pawn.rb. Kudos to the passing tests. I'm giving it a :+1:.

File: app/models/pawn.rb (hunk 3 above)
Comment (leamerluck) [inline app/models/pawn.rb]: Again, at the suggestion of my mentor. I'd like to leave it, but please comment if you are not ok with that.

File: app/models/pawn.rb (hunk 2 above)
Comment (leamerluck) [inline app/models/pawn.rb]: This method was added at the suggestion of my mentor. If there is no objection, I will leave it.

File: app/models/pawn.rb (hunk 4 above)
Comment (leamerluck) [inline app/models/pawn.rb]: Thanks for that tip... I need to do some reading up on Ruby style guide :+1: I will make this change!

**Pull request merged**